search/ranking_info.cpp

```diff
@@ -131,32 +130,78 @@ void PrintParse(ostringstream & oss, array<TokenRange, Model::TYPE_COUNT> const
   oss << "]";
 }
 
-class IsAttraction
+class BaseTypesChecker
 {
-  std::vector<uint32_t> m_sights;
-  uint32_t m_leisure;
+protected:
+  std::vector<uint32_t> m_types;
 
+public:
+  bool operator() (feature::TypesHolder const & th) const
+  {
+    return base::AnyOf(m_types, [&th](uint32_t t) { return th.HasWithSubclass(t); });
+  }
+};
+
+class IsAttraction : public BaseTypesChecker
+{
 public:
   IsAttraction()
   {
     // We have several lists for attractions: short list in search categories for @tourism and long
     // list in ftypes::AttractionsChecker. We have highway-pedestrian, place-square, historic-tomb,
     // landuse-cemetery, amenity-townhall etc in long list and logic of long list is "if this object
     // has high popularity and/or wiki description probably it is attraction". It's better to use
-    // short list here. And leisures too!
+    // short list here.
+    m_types = search::GetCategoryTypes("sights", "en", GetDefaultCategories());
+
+    // Add _attraction_ leisures too!
+    base::StringIL const types[] = {
+      {"leisure", "beach_resort"},
+      {"leisure", "garden"},
+      {"leisure", "landscape_reserve"},
+      {"leisure", "marina"},
+      {"leisure", "nature_reserve"},
+      {"leisure", "park"},
+    };
 
-    m_sights = search::GetCategoryTypes("sights", "en", GetDefaultCategories());
-    m_leisure = classif().GetTypeByPath({"leisure"});
+    Classificator const & c = classif();
+    for (auto const & e : types)
+      m_types.push_back(c.GetTypeByPath(e));
   }
+};
 
-  bool operator() (feature::TypesHolder const & th) const
+class IsShopOrAmenity : public BaseTypesChecker
+{
+public:
+  IsShopOrAmenity()
   {
-    return th.HasWithSubclass(m_leisure) ||
-           base::AnyOf(m_sights, [&th](uint32_t t) { return th.HasWithSubclass(t); });
+    base::StringIL const types[] = {
+      {"shop"},
+
+      // Amenity types are very fragmented, so take only most _interesting_ here.
```

Comment:
Will these amenities have a better rank?

How did you make this list?

Is there a way to build this list dynamically, based on types that were selected/tapped by user on the map?

Reply:
- Yes, better than default.
- Manually looking at the "amenity" hierarchy.
I don't like OSM tagging here, why the hell theatre and cinema are the same amenities like toilet, bench or post_box ?!
- Maybe, not in this scope.